### PR TITLE
Improve Embed Layer Norm Fusion for SQuAD with static input shape 

### PR DESCRIPTION
**Pull request by tianleiwu**
**Description**: SQuAD with static input shape will get position embedding constant folded by ORT. Add a new subgraph pattern in fusion.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

Need this to support SQuAD bert model with static  input shape.

### CODE REVIEW

File: onnxruntime/core/optimizer/embed_layer_norm_fusion.cc [hunk 1]
```diff
@@ -401,8 +543,9 @@ Status EmbedLayerNormFusion::ApplyImpl(Graph& graph, bool& modified, int graph_l
         sg_shape->dim()[1].dim_value() <= 0) {
       continue;
     }
+    auto hidden_size = sg_shape->dim()[1].dim_value();
 
-    // Traceback the SkipLayerNormalization node to find Gather --> Add --> SkipLayerNormalization
+    // Trace back to find Gather --> Add --> SkipLayerNormalization
```
Comment (yufenglee) [inline onnxruntime/core/optimizer/embed_layer_norm_fusion.cc]: >SkipLayerNormalization [](start = 45, length = 22)

Should be LayerNorm

File: onnxruntime/core/optimizer/embed_layer_norm_fusion.cc [hunk 2]
```diff
@@ -419,66 +562,65 @@ Status EmbedLayerNormFusion::ApplyImpl(Graph& graph, bool& modified, int graph_l
...
+      if (input_shape->dim_size() != 2 ||
+          !utils::HasDimValue(input_shape->dim()[0]) ||
+          !utils::HasDimValue(input_shape->dim()[1])) {
+        continue;
+      }
+
+      // check tensor shape [batch_size, sequence_length, hidden_size]
+      int64_t batch_size = input_shape->dim()[0].dim_value();
+      int64_t sequence_length = input_shape->dim()[1].dim_value();
+      int64_t input_size = batch_size * sequence_length;
```
Comment (yufenglee) [inline onnxruntime/core/optimizer/embed_layer_norm_fusion.cc]: so you only fuse when input has static shape here?

File: onnxruntime/core/optimizer/embed_layer_norm_fusion.cc [hunk 3]
```diff
@@ -272,65 +353,126 @@ static bool MatchPositionEmbeddingSubgraph2(
...
+      return nullptr;
+    }
+
+    initializer.set_raw_data(data, element_count * sizeof(MLFloat16));
   }
 
-  // Do not add the gather node since it has been added in another path.
-  matched_edges.push_back(edges[0]);
-  return true;
+  NodeArg& node_arg = graph_utils::AddInitializer(graph, initializer);
```
Comment (yufenglee) [inline onnxruntime/core/optimizer/embed_layer_norm_fusion.cc]: old_intializer and new initializer share same raw data. Is the shared raw buffer still valid after the old initializer is cleaned up?

File: onnxruntime/core/optimizer/embed_layer_norm_fusion.cc (hunk 2 above)
Comment (tianleiwu) [reply onnxruntime/core/optimizer/embed_layer_norm_fusion.cc]: This case only happens for static shape due to constant folding.

File: onnxruntime/core/optimizer/embed_layer_norm_fusion.cc (hunk 3 above)
Comment (tianleiwu) [reply onnxruntime/core/optimizer/embed_layer_norm_fusion.cc]: old_intializer and new one has same life cycle. After graph_utils::AddInitializer, the weight data will be hold by graph.

Comment (yufenglee) [reply onnxruntime/core/optimizer/embed_layer_norm_fusion.cc]: Graph::Resolve calls Graph::CleanUnusedInitializers.

Comment (tianleiwu) [reply onnxruntime/core/optimizer/embed_layer_norm_fusion.cc]: I believe that set_raw_data will get a copy of raw data. You can see that TensorProto has also a function called clear_raw_data(). If it does not own the raw data, there is no need to clear it.

**Pull request merged**